drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,1102 @@
+	 * If Dcache enabled, need to invalidate the cachelines to
+	 * read updated cmpl desc. The cache API also issues dsb.
+	 */
+	dma_mb();
+
+	/* Decode cmpl pkt id to verify */
+	c = (struct cmpl_pkt *)((uintptr_t)ring->cmpl +
+	    PAX_DMA_CMPL_DESC_SIZE * PAX_DMA_CURR_CMPL_IDX(wr_offs));
+
+	LOG_DBG("RING%d WR_PTR:%d opq:%d, rm_status:%x dma_status:%x\n",
+		idx, wr_offs, c->opq, c->rm_status, c->dma_status);
+
+	is_outstanding = sys_read32(RM_RING_REG(pd, idx,
+						RING_NUM_REQ_OUTSTAND));
+	if ((ring->curr.opq != c->opq) && (is_outstanding != 0)) {
+		LOG_ERR("RING%d: pkt id should be %d, rcvd %d outst=%d\n",
+			idx, ring->curr.opq, c->opq, is_outstanding);
+		ret = -EIO;
+	}
+	/* check for completion AE timeout */
+	if (c->rm_status == RM_COMPLETION_AE_TIMEOUT) {
+		LOG_ERR("RING%d WR_PTR:%d rm_status:%x AE Timeout!\n",
+			idx, wr_offs, c->rm_status);
+		/* TBD: Issue full card reset to restore operations */
+		LOG_ERR("Needs Card Reset to recover!\n");
+		ret = -ETIMEDOUT;
+	}
+
+	if (ring->dma_callback) {
+		ring->dma_callback(dev, ring->callback_arg, idx, ret);
+	}
+
+	/* clear total packet count and non header bd count */
+	ring->non_hdr_bd_count = 0;
+	ring->total_pkt_count = 0;
+
+	return ret;
+}
+
+#ifdef CONFIG_DMA_IPROC_PAX_POLL_MODE
+static int peek_ring_cmpl(const struct device *dev,
+			  enum ring_idx idx, uint32_t pl_len)
+{
+	struct dma_iproc_pax_data *pd = PAX_DMA_DEV_DATA(dev);
+	uint32_t wr_offs, rd_offs, timeout = PAX_DMA_MAX_POLL_WAIT;
+	struct dma_iproc_pax_ring_data *ring = &(pd->ring[idx]);
+
+	/* cmpl read offset, unprocessed cmpl location */
+	rd_offs = ring->curr.cmpl_rd_offs;
+
+	/* poll write_ptr until cmpl received for all buffers */
+	do {
+		wr_offs = sys_read32(RM_RING_REG(pd, idx,
+						 RING_CMPL_WRITE_PTR));
+		if (PAX_DMA_GET_CMPL_COUNT(wr_offs, rd_offs) >= pl_len)
+			break;
+		k_busy_wait(1);
+	} while (--timeout);
+
+	if (timeout == 0) {
+		LOG_ERR("RING%d timeout, rcvd %d, expected %d!\n",
+			idx, PAX_DMA_GET_CMPL_COUNT(wr_offs, rd_offs), pl_len);
+		/* More debug info on current dma instance */
+		LOG_ERR("WR_PTR:%x RD_PTR%x\n", wr_offs, rd_offs);
+		return  -ETIMEDOUT;
+	}
+
+	return process_cmpl_event(dev, idx, pl_len);
+}
+#else
+static void rm_isr(void *arg)
+{
+	uint32_t status, err_stat, idx;
+	const struct device *dev = arg;
+	struct dma_iproc_pax_data *pd = PAX_DMA_DEV_DATA(dev);
+
+	err_stat =
+	sys_read32(RM_COMM_REG(pd,
+			       RM_COMM_AE_INTERFACE_GROUP_0_INTERRUPT_MASK));
+	sys_write32(err_stat,
+		    RM_COMM_REG(pd,
+				RM_COMM_AE_INTERFACE_GROUP_0_INTERRUPT_CLEAR));
+
+	/* alert waiting thread to process, for each completed ring */
+	for (idx = PAX_DMA_RING0; idx < PAX_DMA_RINGS_MAX; idx++) {
+		status =
+		sys_read32(RM_RING_REG(pd, idx,
+				       RING_COMPLETION_INTERRUPT_STAT));
+		sys_write32(status,
+			    RM_RING_REG(pd, idx,
+					RING_COMPLETION_INTERRUPT_STAT_CLEAR));
+		if (status & 0x1) {
+			k_sem_give(&pd->ring[idx].alert);
```

Comment:
If the input transfer is divided into multiple packets, the 1st packet will trigger an ISR upon finish. The code here will unblock the start_xfer function with partial data xfer complete. 

If we use extended packet, we can avoid this issue. Let me know if you see a better option.

Reply:
Ring can be made inactive before generating all the packets corresponding to input transfer, Once BD's are populated in the memory- program the "total_pkt_count/header bd's count" count into RING_CMPL_WR_PTR_DDR_CONTROL so that RM generates interrupt only after processing "total_pkt_count/header bds' count".

Reply:
Right. Use set_pkt_count() will also avoid multiple interrupts for 1 xfer, so you don't need use extended packet.